### DISCUSSION TITLE
"Playing" a midi file and calling a method every time a note plays

### DISCUSSION THREAD

**Discussion by ZacharyLaurin** [melanchall/drywetmidi#196] [Q&A] [answered]
I am making a rhythm game in Unity and I was wondering how I could go about "playing" a midi file in sync with an mp3 file and calling a method every time a note plays with it's instrument, velocity, length, etc.? I tried using AtTime() and running it every frame but it only works if the playback time of the mp3 is _exactly_ the same as the time of the note, which is rarely ever the case.

**Comment by melanchall** [comment, marked as answer]
I don't completely understand why you have "playing" word in quotes. You can use `Playback` API of the library and set event handler on notes:

```csharp
_playback = midiFile.GetPlayback();
_playback.EventPlayed += OnEventPlayed;
_playback.NotesPlaybackStarted += OnNotesPlaybackStarted;

// ...

private static void OnEventPlayed(object sender, MidiEventPlayedEventArgs e)
{
    if (e.Event.EventType == MidiEventType.ProgramChange)
        _program = ((ProgramChangeEvent)e.Event).ProgramNumber;
}

private static void OnNotesPlaybackStarted(object sender, NotesEventArgs e)
{
    // do smth with notes, use _program here
}
```

As you can see you need additional handler for MIDI events to store the current program (instrument) and then you can use it in `OnNotesPlaybackStarted`.

More info about `Playback` here: https://melanchall.github.io/drywetmidi/articles/playback/Overview.html.

**Reply by melanchall** [reply]
@ZacharyLaurin Any news?

**Reply by ZacharyLaurin** [reply]
@melanchall Thank you, that works
